### PR TITLE
[BACKLOG-3888] - Moving hdfs steps, job entries to an osgi bundle

### PR DESCRIPTION
**Pull request by brosander**
Build Commands:
==============
Build:
------
1. mvn clean install && mvn site

Cleanup:
-------
mvn build-helper:remove-project-artifact

Changed files:
==============
  api/cluster/src/main/java/org/pentaho/big/data/api/cluster/NamedCluster.java
  assemblies/features/src/main/resources/features.xml
  impl/cluster/src/main/java/org/pentaho/big/data/impl/cluster/NamedClusterImpl.java
  impl/cluster/src/test/java/org/pentaho/big/data/impl/cluster/NamedClusterImplTest.java
  kettle-plugins/common/namedClusterBridge/pom.xml
  kettle-plugins/common/namedClusterBridge/src/main/java/org/pentaho/big/data/plugins/common/ui/named/cluster/bridge/NamedClusterBridgeImpl.java
  kettle-plugins/common/namedClusterBridge/src/main/java/org/pentaho/big/data/plugins/common/ui/named/cluster/bridge/NamedClusterUIFactoryBridgeImpl.java
  kettle-plugins/common/namedClusterBridge/src/test/java/org/pentaho/big/data/plugins/common/ui/named/cluster/bridge/NamedClusterBridgeImplTest.java
  kettle-plugins/hdfs/pom.xml
  kettle-plugins/hdfs/src/main/java/org/pentaho/big/data/kettle/plugins/hdfs/HdfsLifecycleListener.java
  kettle-plugins/hdfs/src/main/java/org/pentaho/big/data/kettle/plugins/hdfs/job/JobEntryHadoopCopyFiles.java
  kettle-plugins/hdfs/src/main/java/org/pentaho/big/data/kettle/plugins/hdfs/job/JobEntryHadoopCopyFilesDialog.java
  kettle-plugins/hdfs/src/main/java/org/pentaho/big/data/kettle/plugins/hdfs/trans/HadoopFileInputDialog.java
  kettle-plugins/hdfs/src/main/java/org/pentaho/big/data/kettle/plugins/hdfs/trans/HadoopFileInputMeta.java
  kettle-plugins/hdfs/src/main/java/org/pentaho/big/data/kettle/plugins/hdfs/trans/HadoopFileOutputDialog.java
  kettle-plugins/hdfs/src/main/java/org/pentaho/big/data/kettle/plugins/hdfs/trans/HadoopFileOutputMeta.java
  kettle-plugins/hdfs/src/main/java/org/pentaho/big/data/kettle/plugins/hdfs/vfs/HadoopVfsConnection.java
  kettle-plugins/hdfs/src/main/java/org/pentaho/big/data/kettle/plugins/hdfs/vfs/HadoopVfsFileChooserDialog.java
  kettle-plugins/hdfs/src/main/java/org/pentaho/big/data/kettle/plugins/hdfs/vfs/MapRFSFileChooserDialog.java
  kettle-plugins/hdfs/src/main/java/org/pentaho/big/data/kettle/plugins/hdfs/vfs/Schemes.java
  kettle-plugins/hdfs/src/main/resources/HDI.png
  kettle-plugins/hdfs/src/main/resources/HDI.svg
  kettle-plugins/hdfs/src/main/resources/HDM.png
  kettle-plugins/hdfs/src/main/resources/HDM.svg
  kettle-plugins/hdfs/src/main/resources/HDO.png
  kettle-plugins/hdfs/src/main/resources/HDO.svg
  kettle-plugins/hdfs/src/main/resources/OSGI-INF/blueprint/blueprint.xml
  kettle-plugins/hdfs/src/main/resources/org/pentaho/big/data/kettle/plugins/hdfs/job/messages/messages_en_US.properties
  kettle-plugins/hdfs/src/main/resources/org/pentaho/big/data/kettle/plugins/hdfs/job/messages/messages_ko_KR.properties
  kettle-plugins/hdfs/src/main/resources/org/pentaho/big/data/kettle/plugins/hdfs/trans/messages/messages_en_US.properties
  kettle-plugins/hdfs/src/main/resources/org/pentaho/big/data/kettle/plugins/hdfs/trans/messages/messages_ko_KR.properties
  kettle-plugins/hdfs/src/main/resources/org/pentaho/big/data/kettle/plugins/hdfs/vfs/messages/messages_en_US.properties
  kettle-plugins/hdfs/src/main/resources/org/pentaho/big/data/kettle/plugins/hdfs/vfs/messages/messages_ko_KR.properties
  kettle-plugins/hdfs/src/test/java/org/pentaho/big/data/kettle/plugins/hdfs/job/JobEntryHadoopCopyFilesTest.java
  kettle-plugins/hdfs/src/test/java/org/pentaho/big/data/kettle/plugins/hdfs/trans/HadoopFileOutputDialogTest.java
  kettle-plugins/hdfs/src/test/java/org/pentaho/big/data/kettle/plugins/hdfs/trans/HadoopFileOutputMetaTest.java
  kettle-plugins/hdfs/src/test/java/org/pentaho/big/data/kettle/plugins/hdfs/vfs/HadoopVfsConnectionTest.java
  kettle-plugins/pom.xml
  legacy/src/main/java/org/pentaho/di/core/hadoop/HadoopSpoonPlugin.java
  legacy/src/main/java/org/pentaho/di/job/entries/hadoopcopyfiles/JobEntryHadoopCopyFiles.java
  legacy/src/main/java/org/pentaho/di/trans/steps/hadoopfileinput/HadoopFileInputMeta.java
  legacy/src/main/java/org/pentaho/di/trans/steps/hadoopfileoutput/HadoopFileOutputMeta.java
  legacy/src/main/java/org/pentaho/di/ui/core/namedcluster/NamedClusterUIFactory.java
  legacy/src/main/java/org/pentaho/di/ui/job/entries/hadoopcopyfiles/JobEntryHadoopCopyFilesDialog.java
  legacy/src/main/java/org/pentaho/di/ui/job/entries/sqoop/AbstractSqoopJobEntryController.java
  legacy/src/main/java/org/pentaho/di/ui/trans/steps/hadoopfileinput/HadoopFileInputDialog.java
  legacy/src/main/java/org/pentaho/di/ui/trans/steps/hadoopfileoutput/HadoopFileOutputDialog.java
  legacy/src/main/java/org/pentaho/di/ui/vfs/HadoopVfsConnection.java
  legacy/src/main/java/org/pentaho/di/ui/vfs/VfsFileChooserHelper.java
  legacy/src/main/java/org/pentaho/di/ui/vfs/hadoopvfsfilechooserdialog/HadoopVfsFileChooserDialog.java
  legacy/src/main/java/org/pentaho/di/ui/vfs/hadoopvfsfilechooserdialog/MapRFSFileChooserDialog.java
  legacy/src/main/resources/HDI.png
  legacy/src/main/resources/HDI.svg
  legacy/src/main/resources/HDM.png
  legacy/src/main/resources/HDM.svg
  legacy/src/main/resources/HDO.png
  legacy/src/main/resources/HDO.svg
  legacy/src/main/resources/org/pentaho/di/job/entries/hadoopcopyfiles/messages/messages_en_US.properties
  legacy/src/main/resources/org/pentaho/di/job/entries/hadoopcopyfiles/messages/messages_ko_KR.properties
  legacy/src/main/resources/org/pentaho/di/trans/steps/hadoopfileinput/messages/messages_en_US.properties
  legacy/src/main/resources/org/pentaho/di/trans/steps/hadoopfileoutput/messages/messages_en_US.properties
  legacy/src/main/resources/org/pentaho/di/trans/steps/hadoopfileoutput/messages/messages_ko_KR.properties
  legacy/src/main/resources/org/pentaho/di/ui/vfs/hadoopvfsfilechooserdialog/messages/messages_en_US.properties
  legacy/src/main/resources/org/pentaho/di/ui/vfs/hadoopvfsfilechooserdialog/messages/messages_ko_KR.properties
  legacy/src/test/java/org/pentaho/di/job/entries/hadoopcopyfiles/JobEntryHadoopCopyFilesTest.java
  legacy/src/test/java/org/pentaho/di/ui/trans/steps/hadoopfileoutput/HadoopFileOutputDialogTest.java
  legacy/src/test/java/org/pentaho/di/ui/vfs/HadoopVfsConnectionTest.java
  legacy/src/test/java/org/pentaho/trans/steps/hadoopfileoutput/HadoopFileOutputTest.java

Checkstyle violations:
======================
See: https://github.com/pentaho/pentaho-coding-standards

File kettle-plugins/hdfs/src/main/java/org/pentaho/big/data/kettle/plugins/hdfs/trans/HadoopFileInputDialog.java
1. Message: File length is 3,051 lines (max allowed is 2,000). at Line: 1 Column: None

Unit test coverage change:
============
org.pentaho.big.data.kettle.plugins.hdfs.vfs.HadoopVfsConnection:
 * branchPercentCoverage :87.50%
 * linePercentCoverage :100.00%
 * methodPercentCoverage :100.00%
 * complexityPercentCoverage :88.89%
 * instructionPercentCoverage :100.00%
    
org.pentaho.big.data.kettle.plugins.hdfs.trans.HadoopFileOutputDialog:
 * branchPercentCoverage :1.29%
 * linePercentCoverage :1.24%
 * methodPercentCoverage :11.11%
 * complexityPercentCoverage :3.09%
 * instructionPercentCoverage :0.71%
    
org.pentaho.big.data.kettle.plugins.hdfs.trans.HadoopFileOutputMeta:
 * linePercentCoverage :20.00%
 * methodPercentCoverage :23.08%
 * complexityPercentCoverage :20.00%
 * instructionPercentCoverage :18.49%
    
org.pentaho.big.data.kettle.plugins.hdfs.job.JobEntryHadoopCopyFiles:
 * branchPercentCoverage :83.33%
 * linePercentCoverage :81.25%
 * methodPercentCoverage :50.00%
 * complexityPercentCoverage :55.56%
 * instructionPercentCoverage :83.64%
    
org.pentaho.di.ui.job.entries.sqoop.AbstractSqoopJobEntryController:
 * instructionPercentCoverageChange : +0.39%
 * branchPercentCoverageChange : +1.04%
 * linePercentCoverageChange : +0.56%
 * complexityPercentCoverageChange : +0.49%
    
org.pentaho.di.ui.vfs.VfsFileChooserHelper:
 * instructionPercentCoverageChange :**-9.57%**
 * branchPercentCoverageChange :**-2.68%**
 * linePercentCoverageChange :**-6.65%**
 * complexityPercentCoverageChange :**-0.37%**
 * methodPercentCoverageChange : +1.23%
    
org.pentaho.big.data.plugins.common.ui.named.cluster.bridge.NamedClusterBridgeImpl:
 * branchPercentCoverageChange :**-12.50%**
 * complexityPercentCoverageChange :**-2.04%**
    
org.pentaho.big.data.impl.cluster.NamedClusterImpl:
 * instructionPercentCoverageChange :**-24.24%**
 * branchPercentCoverageChange :**-32.94%**
 * linePercentCoverageChange :**-20.62%**
 * complexityPercentCoverageChange :**-26.65%**
 * methodPercentCoverageChange :**-4.26%**

### CODE REVIEW

File: kettle-plugins/common/namedClusterBridge/src/main/java/org/pentaho/big/data/plugins/common/ui/named/cluster/bridge/NamedClusterBridgeImpl.java [hunk 1]
```diff
@@ -189,6 +202,21 @@ public NamedCluster clone() {
...
+  public String processURLsubstitution( String incomingURL, IMetaStore metastore, VariableSpace variableSpace ) {
+    if ( isMapr() ) {
+      String url = namedClusterManager
+        .processURLsubstitution( getName(), incomingURL, HadoopSpoonPlugin.MAPRFS_SCHEME, metastore, variableSpace );
+      if ( url != null && !url.startsWith( HadoopSpoonPlugin.MAPRFS_SCHEME ) ) {
+        url = HadoopSpoonPlugin.MAPRFS_SCHEME + "://" + url;
+      }
+      return url;
+    } else {
+      return namedClusterManager
```
Comment (mkambol) [inline kettle-plugins/common/namedClusterBridge/src/main/java/org/pentaho/big/data/plugins/common/ui/named/cluster/bridge/NamedClusterBridgeImpl.java]: The NamedClusterManager at this line is the legacy NamedClusterManager-- is this intentional?  I would have expected it to leverage the processURLsubstitution from the new Impl class above.

**Comment by brosander** [comment]
@mkambol the bridged impl is there to facilitate interop between OSGi and the legacy plugin.

This implementation is using logic living in the legacy plugin to provide an implementation of the new NamedCluster interface.  This allows the legacy plugin to use the named cluster testing functionality provided by OSGi.

**Comment by mkambol** [comment]
The code duplication between the new NamedClusterImpl and the legacy NamedClusterManager makes me itch.  When does that go away?

**Comment by brosander** [comment]
@mkambol when we finish osgiifying all the services and I agree

**Pull request merged**